### PR TITLE
#13181 - Enhance National Health ID Search Functionality Across Modules and Address Duplication Issues

### LINKED ISSUE #13181: Enhance National Health ID Search Functionality Across Modules and Address Duplication Issues [3]
### Problem Description

This issue is related to [Search by NationalID is not consistently working in all locations · Issue #13029](https://github.com/SORMAS-Foundation/SORMAS-Project/issues/13029).

Problem Summary:
Duplication Issue: Occasionally, two different individuals associated with the same National Health ID are created. It was observed this while manually processing a lab message. The duplication may stem from variations in the first and last names of these individuals.

Search Limitation in Person Module: When attempting to merge these records in the Person module, you cannot search by the National Health ID. Instead, the search results display a completely different individual, which suggests that the National Health ID search functionality may not have been included in the scope of issue #13029.

### Proposed Change

1. Enable Comprehensive National Health ID Search Across Modules:

Implement a search feature extension allowing users to search by National Health ID across the Case, Person, Contacts Sample, and Message modules. This would allow consistent and accurate retrieval of records associated with a particular National Health ID.
In the Person module, specifically, enhance the search functionality to prioritize National Health ID results, reducing confusion when merging records.

2. Prevent Duplicate Record Creation Based on National Health ID:

Implement validation logic to check for existing National Health ID entries when processing new records (e.g., lab messages) to prevent duplicate entries due to minor name variations.
Display a warning or notification if a duplicate National Health ID is detected, prompting users to review and confirm before creating a new record.

### Added Value/Benefit

- Enhanced Usability and User Confidence
- Reduced Risk of Data Duplication Issues in Reporting and Analysis
- Improved Data Accuracy and Consistency

### Acceptance Criteria

Enhanced National Health ID Search Functionality Across Modules

- [ ]  Users can search by National Health ID across the Case (exists), Person, Contacts(exists), Sample, and Message modules.
- [ ]  The search returns only relevant records associated with the specified National Health ID, reducing mismatched search results.
- [ ] In the Person module, the search prioritizes results based on the National Health ID when merging records.
- [ ] Display a warning or notification if a duplicate National Health ID is detected

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #13181

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13201

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13201

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13201

File: sormas-api/src/main/resources/strings.properties [hunk 1]
```diff
@@ -697,6 +697,7 @@ headingShowExternalMessage = Message
 headingSelfReportSideComponent = Self reports
 headingSignsAndSymptoms = Clinical Signs and Symptoms
 headingSimilarImmunization = Similar immunizaton
+headingSimilarPerson = Similar persons found based on person introduced data
```
Comment (leventegal-she) [inline sormas-api/src/main/resources/strings.properties:700]: This sentence doesn't seem correct, I think we should be precise and say There are some other persons with ht e same national health id.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CaseController.java [hunk 2]
```diff
@@ -151,6 +152,9 @@
 
 public class CaseController {
 
+	CommitDiscardWrapperComponent<CaseCreateForm> caseCreateComponent;
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CaseController.java:155]: missing access modifiers

File: sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactController.java [hunk 3]
```diff
@@ -110,6 +111,8 @@
 public class ContactController {
 
 	protected final Logger logger = LoggerFactory.getLogger(getClass());
+	boolean contactSaveTriggered;
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactController.java:114]: missing access modifiers

File: sormas-ui/src/main/java/de/symeda/sormas/ui/immunization/ImmunizationController.java [hunk 4]
```diff
@@ -40,14 +41,17 @@
 
 public class ImmunizationController {
 
+	boolean immunizationSaveTriggered;
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/immunization/ImmunizationController.java:44]: same

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java [hunk 5]
```diff
@@ -43,14 +45,17 @@
 
 public class TravelEntryController {
 
+	boolean travelEntrySaveTriggered;
+	CommitDiscardWrapperComponent<TravelEntryCreateForm> travelEntryCreateComponent;
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java:49]: same

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditForm.java
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/AbstractEditForm.java:1]: please revert the changes in this file

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/components/TextFieldCustom.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package de.symeda.sormas.ui.utils.components;
+
+import java.lang.reflect.Method;
+
+import com.vaadin.shared.Registration;
+import com.vaadin.v7.data.Property;
+import com.vaadin.v7.ui.TextField;
+
+public class TextFieldCustom extends TextField {
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/components/TextFieldCustom.java:9]: Please rename to SormasTextField

**Review by leventegal-she — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/resources/strings.properties (hunk 1 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:700]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CaseController.java (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CaseController.java:155]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactController.java (hunk 3 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactController.java:114]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/immunization/ImmunizationController.java (hunk 4 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/immunization/ImmunizationController.java:44]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java (hunk 5 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java:49]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/components/TextFieldCustom.java (hunk 6 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/components/TextFieldCustom.java:9]: renamed

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13201

**Review by leventegal-she — APPROVED** [review]
